File: mobile/lib/services/device_identity.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:sodium_libs/sodium_libs.dart';

/// This phone's long-lived ed25519 identity. Generated once on first launch
/// and persisted in secure storage; never re-rolled (re-pairing per desktop
/// is what changes; the device identity is stable).
class DeviceIdentity {
  static final DeviceIdentity instance = DeviceIdentity._();
  DeviceIdentity._();

  static const _storage = FlutterSecureStorage(
    aOptions: AndroidOptions(resetOnError: true, encryptedSharedPreferences: true),
  );

  String deviceId = '';
  Uint8List publicKey = Uint8List(0);
  Uint8List _privateKey = Uint8List(0); // 64-byte libsodium ed25519 secret-key form

  Sodium? _sodium;
  Future<Sodium> _ready() async => _sodium ??= await SodiumInit.init();

  /// Read or generate the identity. Idempotent — calling it twice is safe and
  /// the second call short-circuits.
  Future<void> load() async {
    if (deviceId.isNotEmpty) return;
    final sodium = await _ready();

    final existingId = await _safeRead('device_id');
    final pub = await _safeRead('device_public_key');
    final priv = await _safeRead('device_private_key');
    if (existingId != null && pub != null && priv != null) {
      deviceId = existingId;
      publicKey = base64Decode(pub);
      _privateKey = base64Decode(priv);
      return;
    }

    // First launch — generate.
    final kp = sodium.crypto.sign.keyPair();
    deviceId =
        'clippy-phone-${sodium.randombytes.buf(8).map((b) => b.toRadixString(16).padLeft(2, '0')).join()}';
    publicKey = Uint8List.fromList(kp.publicKey);
    _privateKey = kp.secretKey.extractBytes();
    kp.secretKey.dispose();

    try {
      await _storage.write(key: 'device_id', value: deviceId);
      await _storage.write(key: 'device_public_key', value: base64Encode(publicKey));
      await _storage.write(key: 'device_private_key', value: base64Encode(_privateKey));
    } catch (_) {}
  }

  /// Sign `message` with this device's private key (ed25519 detached).
  Future<Uint8List> sign(Uint8List message) async {
    final sodium = await _ready();
    final sk = SecureKey.fromList(sodium, _privateKey);
    try {
      return sodium.crypto.sign.detached(message: message, secretKey: sk);
    } finally {
      sk.dispose();
    }
  }

  Future<String?> _safeRead(String key) async {
    try {
      return await _storage.read(key: key);
    } catch (_) {
      return null;
    }
  }
}

File: desktop/electron/device-identity.ts
import sodium from 'libsodium-wrappers';
import type { Db } from './db';

/** This desktop's long-lived ed25519 keypair + stable device_id.
 *  Generated lazily on first request and persisted in `device_identity`. */
export interface DeviceIdentity {
  deviceId: string;
  publicKey: Uint8Array;   // 32 bytes
  privateKey: Uint8Array;  // 64 bytes (libsodium ed25519 secret-key form)
}

let cached: DeviceIdentity | null = null;

export async function getDeviceIdentity(db: Db): Promise<DeviceIdentity> {
  if (cached) return cached;
  await sodium.ready;

  const row = db
    .raw()
    .prepare(
      "SELECT device_id, public_key, private_key FROM device_identity WHERE k = 'self'"
    )
    .get() as { device_id: string; public_key: Buffer; private_key: Buffer } | undefined;

  if (row) {
    cached = {
      deviceId: row.device_id,
      publicKey: new Uint8Array(row.public_key),
      privateKey: new Uint8Array(row.private_key),
    };
    return cached;
  }

  // First launch on this desktop — generate + persist.
  const kp = sodium.crypto_sign_keypair();
  const deviceId = `clippy-desktop-${sodium.to_hex(sodium.randombytes_buf(8))}`;
  db.raw()
    .prepare(
      `INSERT INTO device_identity(k, device_id, public_key, private_key, created_at)
       VALUES ('self', ?, ?, ?, ?)`
    )
    .run(deviceId, Buffer.from(kp.publicKey), Buffer.from(kp.privateKey), Date.now());

  cached = { deviceId, publicKey: kp.publicKey, privateKey: kp.privateKey };
  return cached;
}

/** Sign `message` with this device's private key (ed25519 detached). */
export async function signWithDeviceKey(
  db: Db,
  message: Uint8Array
): Promise<Uint8Array> {
  const id = await getDeviceIdentity(db);
  await sodium.ready;
  return sodium.crypto_sign_detached(message, id.privateKey);
}
